<launch>

    <!-- ==================== Experiment Args ==================== -->
    <arg name="real_experiment" default="false"/>
    <arg name="has_simulation" default="true"/>
    <arg name="slip_detection" default="false"/>
    <arg name="num_imus" default="0"/>
    <arg name="adapt_b" default="false"/>
    <arg name="inf" default="false"/>
    <arg name="wname" default="earth"/>
    <arg name="tips" default="false"/>
    <arg name="walk" default="false"/>

    <group ns="/" unless="$(arg real_experiment)">
        <!-- ==================== GAZEBO WORLD ==================== -->
        <arg name="paused" default="true"/>
        <arg name="use_sim_time" default="true"/>
        <arg name="gui" default="true"/>
        <arg name="headless" default="false"/>
        <arg name="debug" default="false"/>
        
        <include file="$(find gazebo_ros)/launch/empty_world.launch">
            <arg name="world_name" value="$(find maestro)/worlds/$(arg wname).world"/> 
            <arg name="debug" value="$(arg debug)"/>
            <arg name="gui" value="$(arg gui)"/>
            <arg name="paused" value="$(arg paused)"/>
            <arg name="use_sim_time" value="$(arg use_sim_time)"/>
            <arg name="headless" value="$(arg headless)"/>
        </include>
    </group>

    <group ns="/" if="$(arg real_experiment)">
        <include file="$(find unitree_legged_real)/launch/real.launch">
            <arg name="ctrl_level" value="lowlevel"/> 
        </include>
    </group>

    <group ns="/" >

         <!-- ==================== PARAMETERS ==================== -->
         <!-- REAL Experiment -->
        <!-- <param name="kp" value="2400.0"/>
        <param name="kv" value="280.0"/>
        <param name="ko" value="15.0"/>
        <param name="alpha" value="1000.0"/> -->
<!-- 
        <param name="pbc_x" value="0.001">
        <param name="pbc_y" value="0.001">
        <param name="pbc_z" value="0.0"> -->

         <!-- SIMULATION Experiment -->
        <!-- <param name="kp" value="3000.0"/>
        <param name="kv" value="550.0"/> 
        <param name="ko" value="15.0"/> -->

         <!-- SIMULATION Experiment Walk-->
        <param name="kp" value="1300.0"/>   <!--value="2400.0"-->
        <param name="kv" value="100.0"/>    <!-- value="200.0"--> 
        <param name="ko" value="15.0"/>    <!-- value="150.0"-->
        <param name="ki" value="50.0"/>    <!-- value="350.0"-->

        <param name="alpha" value="150.0"/>  <!-- 150-->

        <param name="pbc_x" value="0.0025"/>  <!-- 0.0025-->
        <param name="pbc_y" value="0.001"/> <!-- 0.001-->
        <param name="pbc_z" value="0.0"/> <!-- 150-->

        <param name="modelstate_id" value="5"/>
        <param name="swingL_id" value="0"/>

        <param name="slip_detection" value="$(arg slip_detection)"/>
        <param name="num_imus" value="$(arg num_imus)"/>
        <param name="adapt_b" value="$(arg adapt_b)"/>
        <param name="inf" value="$(arg inf)"/>
        <param name="tips" value="$(arg tips)"/>
        <param name="walk" value="$(arg walk)"/>

        <param name="robot_name" value="go1"/>
        <param name="real_experiment" value="$(arg real_experiment)"/>
        <param name="has_simulation" value="$(arg has_simulation)"/>
        <param name="urdf_file_path" value="$(find maestro)/go1_description/urdf/go1_slip_stack.urdf"/> <!--go1_slip_stack -->
        <param name="main_path" value="$(find maestro)/CSV/"/>

        <!-- ==================== ROBOT XACRO ==================== -->
        <!-- Debug mode will hung up the robot, use "true" or "false" to switch it. -->
        <arg name="user_debug" default="false"/>
        <param name="robot_description" textfile="$(find maestro)/go1_description/urdf/go1_slip_stack.urdf"/>

        <!-- ==================== GAZEBO ==================== -->
        <node pkg="gazebo_ros" type="spawn_model" name="urdf_spawner" respawn="false" output="screen"
            args="-urdf -x -0.10 -y -0.0 -z 0.6 -model go1_gazebo -param robot_description -unpause"
            unless="$(arg real_experiment)" />
    
        <!-- ==================== YAML ==================== -->
        <!-- Load controller name from YAML file to parameter server -->
        <rosparam file="$(find maestro)/go1_description/config/joint_names.yaml" command="load"/>
        <!-- Load joint controller configurations from YAML file to parameter server -->
        <rosparam file="$(find maestro)/go1_description/config/robot_control.yaml" command="load"/>

        <!-- ==================== JOINT n' ROBOT STATE PUBLISHER ==================== -->        

        <!-- <node pkg="joint_state_publisher_gui" type="joint_state_publisher_gui" name="joint_state_publisher_gui">
            <param name="use_gui" value="TRUE"/>
        </node> -->

            <!-- load the controllers -->
        <node pkg="controller_manager" type="spawner" name="controller_spawner" respawn="false"
          output="screen" ns="/go1_gazebo" args="joint_state_controller
          FL_hip_controller FL_thigh_controller FL_calf_controller
          FR_hip_controller FR_thigh_controller FR_calf_controller
          RL_hip_controller RL_thigh_controller RL_calf_controller
          RR_hip_controller RR_thigh_controller RR_calf_controller "
          unless="$(arg real_experiment)"  />
          
        <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher" 
            unless="$(arg real_experiment)" >
            <remap from="/joint_states" to="/go1_gazebo/joint_states"/>
        </node>

        <!-- ==================== RVIZ ==================== -->
        <!-- <node pkg="rviz" type="rviz" name="rviz" respawn="false" output="screen"
            args="-d $(find maestro)/rviz/go1_maestro.rviz"
            unless="$(arg real_experiment)"  />  -->
           
        <node pkg="maestro"  name="maestro_gazebo_interpreter" type="maestro_gazebo_interpreter"  
            output="screen" unless="$(arg real_experiment)"/> 

    </group>

</launch>